src/test/HostActivationTests/FrameworkResolution/RollForwardMultipleFrameworks.cs

```diff
@@ -0,0 +1,544 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.DotNet.Cli.Build;
+using Microsoft.DotNet.Cli.Build.Framework;
+using System;
+using Xunit;
+
+namespace Microsoft.DotNet.CoreSetup.Test.HostActivation.FrameworkResolution
+{
+    public class RollForwardMultipleFrameworks :
+        FrameworkResolutionBase,
+        IClassFixture<RollForwardMultipleFrameworks.SharedTestState>
+    {
+        private const string MiddleWare = "MiddleWare";
+        private const string AnotherMiddleWare = "AnotherMiddleWare";
+        private const string HighWare = "HighWare";
+
+        private SharedTestState SharedState { get; }
+
+        public RollForwardMultipleFrameworks(SharedTestState sharedState)
+        {
+            SharedState = sharedState;
+        }
+
+        // Soft roll forward from the inner framework reference [specified] to app's 5.1.1 (defaults)
+        [Theory]
+        [InlineData("5.0.0", Constants.RollForwardSetting.Disable, null)]
+        [InlineData("5.1.3", Constants.RollForwardSetting.Disable, "5.1.3")]
+        [InlineData("5.0.0", Constants.RollForwardSetting.LatestPatch, null)]
+        [InlineData("5.1.0", Constants.RollForwardSetting.LatestPatch, "5.1.3")]
+        [InlineData("5.0.0", null, "5.1.3")]
+        [InlineData("5.0.0", Constants.RollForwardSetting.Minor, "5.1.3")]
+        [InlineData("5.0.0", Constants.RollForwardSetting.LatestMinor, "5.1.3")] // The app reference which is Minor wins
+        [InlineData("1.0.0", Constants.RollForwardSetting.Minor, null)]
+        [InlineData("1.0.0", Constants.RollForwardSetting.Major, "5.1.3")]       // The app reference which is Minor wins
+        [InlineData("1.0.0", Constants.RollForwardSetting.LatestMajor, "5.1.3")] // The app reference which is Minor wins
+        public void SoftRollForward_InnerFrameworkReference_ToHigher(
+            string versionReference,
+            string rollForward,
+            string resolvedFramework)
+        {
+            RunTest(
+                runtimeConfig => runtimeConfig
+                    .WithFramework(MiddleWare, "2.1.0")
+                    .WithFramework(MicrosoftNETCoreApp, "5.1.1"),
+                dotnetCustomizer => dotnetCustomizer.Framework(MiddleWare).RuntimeConfig(runtimeConfig =>
+                    runtimeConfig.GetFramework(MicrosoftNETCoreApp)
+                        .WithRollForward(rollForward)
+                        .Version = versionReference),
+                resolvedFramework,
+                commandResult => commandResult.Should().Fail().And.FailedToSoftRollForward(MicrosoftNETCoreApp, versionReference, "5.1.1"));
```

Comment:
Maybe these `Should().Fail()` fails are just bugs.  Some of the tests seem like they should pass.....  Or perhaps this is because these are not really well written fluent assertions.

Reply:
This is just rather bad design of parameters... I'll see about fixing this.
The function looks at the `commandResult` action only if the `resolvedFramework` is null, which makes the call site rather confusing.

Reply:
Even changing the `commandResult` name could help improve comprehension.  i.e. `commandResultWhenResolvedIsNull`

Reply:
Fixed the design... should be much more readable now.